src/charm.py

```diff
@@ -177,116 +173,55 @@ def _on_get_certificate_action(self, event: ActionEvent) -> None:
             event: Juju event
         """
         hostname = event.params["hostname"]
-        tls_information = TLSInformation.from_charm(self)
-
-        tls_rel_data = tls_information.tls_requirer_integration.data[self.app]
-        if any(
-            not tls_rel_data.get(key)
-            for key in [f"certificate-{hostname}", f"ca-{hostname}", f"chain-{hostname}"]
-        ):
-            event.fail("Missing or incomplete certificate data")
-            return
-
-        event.set_results(
-            {
-                f"certificate-{hostname}": tls_rel_data[f"certificate-{hostname}"],
-                f"ca-{hostname}": tls_rel_data[f"ca-{hostname}"],
-                f"chain-{hostname}": tls_rel_data[f"chain-{hostname}"],
-            }
-        )
+        TLSInformation.from_charm(self, self.certificates)
+
+        for cert in self.certificates.get_provider_certificates():
+            if get_hostname_from_cert(cert.certificate) == hostname:
+                event.set_results(
+                    {
+                        "certificate": cert.certificate,
+                        "ca": cert.ca,
+                        "chain": cert.chain_as_pem(),
+                    }
+                )
+                return
+
+        event.fail(f"Missing or incomplete certificate data for {hostname}")
 
     @validate_config_and_integration(defer=True)
     def _on_certificates_relation_created(self, _: RelationCreatedEvent) -> None:
         """Handle the TLS Certificate relation created event."""
-        tls_information = TLSInformation.from_charm(self)
-
+        TLSInformation.from_charm(self, self.certificates)
         client = _get_client(field_manager=self.app.name, namespace=self.model.name)
         config = CharmConfig.from_charm(self, client)
-
-        self._tls.certificate_relation_created(
-            config.external_hostname, tls_information.tls_requirer_integration
-        )
+        self._tls.certificate_relation_created(config.external_hostname)
 
     @validate_config_and_integration(defer=True)
     def _on_certificates_relation_joined(self, _: RelationJoinedEvent) -> None:
         """Handle the TLS Certificate relation joined event."""
-        tls_information = TLSInformation.from_charm(self)
-
+        TLSInformation.from_charm(self, self.certificates)
```

Comment:
same here

Reply:
we are replying on the `from_charm` method to check for the existence of the `certificates` integration and raises an exception which is handled by the `@validate_config_and_integration` decorator. I've added another class method called `validate` and used it here to be more explicit on what we're trying to do.